### PR TITLE
O3-1335: Implement the setup subscription functionality in OCL module

### PR DESCRIPTION
**Pull request by Piumal1999**
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [x] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

Created a micro frontend component for managing an OCL subscription. This UI allows the user to,
1. Save a subscription
2. View the saved subscription
3. Edit the saved subscription
4. Unsubscribe

## Screenshots

image

https://user-images.githubusercontent.com/27498587/175507750-e966192d-c3d9-4290-b1f3-88fe908b862a.mov

## Issue

https://issues.openmrs.org/browse/O3-1335

## Other

*None.*

### CODE REVIEW

**Comment by Piumal1999** [comment]
cc: @ibacher @hadijahkyampeire

File: packages/esm-admin-openconceptlab-app/src/subscription/subscription.component.tsx [hunk 1]
```diff
@@ -36,11 +61,18 @@ const Subscription: React.FC = () => {
...
+          url: subscriptionUrl,
+          token: token,
+          validationType: validationType,
+          subscribedToSnapshot: isSubscribedToSnapshot,
+        },
         abortController,
       );
 
       if (response.ok) {
+        mutate('/ws/rest/v1/openconceptlab/subscription?v=full');
```
Comment (ibacher) [inline packages/esm-admin-openconceptlab-app/src/subscription/subscription.component.tsx:75]: So I think the proper way to do things here would be to do something like:

```ts
const updatedSubscription = {
	...subscription,
  url: subscriptionUrl,
  token: token,
  validationType: validationType,
  subscribedToSnapshot: isSubscribedToSnapshot,
};

mutate("/ws/rest/v1/openconceptlab/subscription?v=full", updatedSubscription, false);

const response = await updateSubscription(updatedSubscription);

mutate("/ws/rest/v1/openconceptlab/subscription?v=full");

if (response.ok) {
   // as before
}
```

**Review by ibacher — COMMENTED** [review]
Thanks @Piumal1999! This mostly LGTM, though I've noted a few things I think should be changed.

File: packages/esm-admin-openconceptlab-app/src/subscription/subscription.component.tsx [hunk 2]
```diff
@@ -56,59 +88,135 @@ const Subscription: React.FC = () => {
 
       return () => abortController.abort();
     },
-    [subscriptionUrl, token],
+    [subscriptionUrl, token, validationType, isSubscribedToSnapshot, t, subscription, mutate],
   );
 
+  const handleCancel = useCallback(() => {
+    setSubscriptionUrl(subscription?.url || '');
```
[the remark this answers is not in the document]
Comment (Piumal1999) [reply packages/esm-admin-openconceptlab-app/src/subscription/subscription.component.tsx:95]: Yeah, I thought the same. But when there's no existing subscription, it doesn't work as intended.

Preview:

https://user-images.githubusercontent.com/27498587/175771512-3eeb3eb7-bc69-4c84-8048-d66d72944f65.mov

File: packages/esm-admin-openconceptlab-app/src/subscription/subscription.resource.ts [hunk 3]
```diff
@@ -24,6 +24,20 @@ export async function updateSubscription(subscription: Subscription, abortContro
...
+  });
+}
+
+export async function deleteSubscription(subscription: Subscription, abortController?: AbortController) {
+  const url = `/ws/rest/v1/openconceptlab/subscription/${subscription.uuid}`;
+  return openmrsFetch<Subscription>(url, {
+    method: 'DELETE',
+    headers: {
+      'Content-Type': 'application/json',
+    },
```
[the remark this answers is not in the document]
Comment (Piumal1999) [reply packages/esm-admin-openconceptlab-app/src/subscription/subscription.resource.ts]: Oh yes. Fixed with de2ccd3

File: packages/esm-admin-openconceptlab-app/src/subscription/subscription.component.tsx (hunk 1 above)
Comment (Piumal1999) [reply packages/esm-admin-openconceptlab-app/src/subscription/subscription.component.tsx]: Fixed with de2ccd3

**Comment by Piumal1999** [comment]
Hi @ibacher, I updated the pull request with the changes you requested. Could you please review again?

**Pull request merged**